### PR TITLE
Anchorage 4.0

### PR DESCRIPTION
**Pull request by jvisenti**
Anchorage 4 adds a new batching API for creating multiple constraints, with the ability to return inactive constraints if you need them. It also improves the API in many places, restructures the code to be more maintainable, and drops Swift 2.3 support. See the ReadMe for a full rundown of the current feature set.

# New Features
- Batching: any constraints created in the batch will not be activated until the whole batch returns. Optionally, you can write `batch(active: false) { ... }` to get back inactive constraints with Auto Layout's batch-activation mechanism.
- Added `sizeAnchors`: you can now pin a view's size to a literal `CGSize` or another view's size anchors in a single line of code.
- Added support for using `(UI/NS)EdgeInsets` when constraining edge anchors; useful for margins when not using `layoutMarginsGuide`.

# Improvements
- Expanded and improved the use of layout priority shorthand enum from #33. This is now the preferred way to adjust priority in Anchorage.
- Added tests

# Breaking Changes
- dropped Swift 2.3 support
- minor changes that may make some priority-related code fail to compile. This code should be converted to the much more concise new shorthand syntax; see the ReadMe for details.

# Bug fixes
- Fixed copy/paste errors that resulted in incorrect constraints being created (https://github.com/Raizlabs/Anchorage/pull/37/commits/9891570de90226eb3bd2c09c48c30b4ffc0e19d9)

### CODE REVIEW

File: Source/Anchorage.swift [hunk 1]
```diff
@@ -65,12 +65,22 @@ public protocol LayoutAxisType {}
...
-#if swift(>=3.0)
+extension BinaryFloatingPoint {
+
+    public init<T: BinaryFloatingPoint>(_ bfp: T) {
+        self.init(
+            sign: bfp.sign,
+            exponentBitPattern: RawExponent(bfp.exponentBitPattern.toUIntMax()),
+            significandBitPattern: RawSignificand(bfp.significandBitPattern.toUIntMax())
+        )
+    }
```
Comment (jvisenti) [inline Source/Anchorage.swift]: Note: This is currently blatantly incorrect, I'm working on a fix now

**Comment by ZevEisenberg** [comment]
Added better debug description because of `NSLayoutAttribute`'s annoying property where its `String(describing:)` is always just `"NSLayoutAttribute"`, as demonstrated in this failing build: https://circleci.com/gh/Raizlabs/Anchorage/2

**Comment by ZevEisenberg** [comment]
Compare to much better reporting in this build: https://circleci.com/gh/Raizlabs/Anchorage/3

File: README.md [hunk 2]
```diff
@@ -88,6 +90,19 @@ let topConstraint = (imageView.topAnchor == container.topAnchor)
 let edgeConstraints = (button.edgeAnchors == container.edgeAnchors).all
 ```
 
+## Creating Inactive Constraints
+
+By default, Anchorage returns active layout constraints. If you'd rather return inactive constraints for use with the [`NSLayoutConstraint.activate(_:)` method](https://developer.apple.com/reference/uikit/nslayoutconstraint/1526955-activate) for performance reasons, you can do it like this:
+
+```swift
+let constraints = Anchorage.inactiveConstraints {
```
Comment (KingOfBrian) [inline README.md]: I think a version of this that returns active constraints would also be helpful, both to activate a group at once for performance reasons and to just collect a set of constraints to manage as a group.

File: Source/AnchorGroupProvider.swift [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    #if os(macOS)
+    public var verticalAnchors: AnchorPair<NSLayoutYAxisAnchor, NSLayoutYAxisAnchor> {
+        return view.verticalAnchors
+    }
+    #else
+    public var verticalAnchors: AnchorPair<NSLayoutYAxisAnchor, NSLayoutYAxisAnchor> {
+        return AnchorPair(first: topLayoutGuide.bottomAnchor, second: bottomLayoutGuide.topAnchor)
+    }
+    #endif
```
Comment (KingOfBrian) [inline Source/AnchorGroupProvider.swift]: I think you can move the `#if` inside the var for a little consolidation.

File: Source/Anchorage.swift [hunk 4]
```diff
@@ -240,420 +237,124 @@ precedencegroup PriorityPrecedence {
...
-    init(horizontal: AnchorPair<NSLayoutXAxisAnchor, NSLayoutXAxisAnchor>, vertical: AnchorPair<NSLayoutYAxisAnchor, NSLayoutYAxisAnchor>) {
-        self.horizontalAnchors = horizontal
-        self.verticalAnchors = vertical
+/// Any Anchorage constraints created inside the passed closure are returned in the array.
+///
+/// - Precondition: Can't be called inside or simultaneously with another batch. Batches cannot be nested.
+/// - Parameter active: Whether the created constraints should be active when they are returned.
+/// - Parameter closure: A closure that runs some Anchorage expressions.
+/// - Returns: An array of new `NSLayoutConstraint`s.
+@discardableResult public func batch(active: Bool, closure: () -> Void) -> [NSLayoutConstraint] {
```
Comment (KingOfBrian) [inline Source/Anchorage.swift]: It might be nice to default `active = true`

File: Source/Compatability.swift [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                pattern.sig = sig >> UIntMax(bitPattern: IntMax(T.significandBitCount - Self.significandBitCount))
+            }
+        }
+
+        self.init(
+            sign: value.sign,
+            exponentBitPattern: RawExponent(pattern.exp),
+            significandBitPattern: RawSignificand(pattern.sig)
+        )
+    }
```
Comment (KingOfBrian) [inline Source/Compatability.swift]: This looks sweet -- I think @robcadwallader will find it interesting too

File: Source/Internal.swift [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    internal typealias LayoutGuide = NSLayoutGuide
+
+    internal let LayoutPriorityRequired = NSLayoutPriorityRequired
+    internal let LayoutPriorityHigh = NSLayoutPriorityDefaultHigh
+    internal let LayoutPriorityLow = NSLayoutPriorityDefaultLow
+    internal let LayoutPriorityFittingSize = NSLayoutPriorityFittingSizeCompression
+#else
+    import UIKit
+
+    internal typealias View = UIView
```
Comment (KingOfBrian) [inline Source/Internal.swift]: :golf: 👏

File: Source/Compatability.swift (hunk 5 above)
Comment (robcadwallader) [inline Source/Compatability.swift]: Yeah, this gives me an idea for some integer improvements

**Comment by ZevEisenberg** [comment]
Here's my take at the release notes:

<hr />

Anchorage 4 adds a new batching API for creating multiple constraints, with the ability to return inactive constraints if you need them. It also improves the API in many places, restructures the code to be more maintainable, and drops Swift 2.3 support. See the ReadMe for a full rundown of the current feature set.

# New Features
- Batching: any constraints created in the batch will not be activated until the whole batch returns. Optionally, you can write `batch(active: false) { ... }` to get back inactive constraints with Auto Layout's batch-activation mechanism.
- Added `sizeAnchors`: you can now pin a view's size to a literal `CGSize` or another view's size anchors in a single line of code.
- Added support for using `(UI/NS)EdgeInsets` when constraining edge anchors; useful for margins when not using `layoutMarginsGuide`.

# Improvements
- Expanded and improved the use of layout priority shorthand enum from #33. This is now the preferred way to adjust priority in Anchorage.
- Added tests

# Breaking Changes
- dropped Swift 2.3 support
- minor changes that may make some priority-related code fail to compile. This code should be converted to the much more concise new shorthand syntax; see the ReadMe for details.

# Bug fixes
- Fixed copy/paste errors that resulted in incorrect constraints being created (https://github.com/Raizlabs/Anchorage/pull/37/commits/9891570de90226eb3bd2c09c48c30b4ffc0e19d9)

**Pull request merged**